src/core/reactor_backend.cc

```diff
@@ -1136,6 +1136,10 @@ try_create_uring(unsigned queue_len, bool throw_on_error) {
             IORING_OP_READV,
             IORING_OP_WRITEV,
             IORING_OP_FSYNC,
+            IORING_OP_RECV,
+            IORING_OP_RECVMSG,
+            IORING_OP_SEND,
+            IORING_OP_SENDMSG,
```

Comment:
But the kernel we run on can be older.

Reply:
ahh, right! i checked the linux source tree: 
- readv, writev were introduced in linux v5.1: https://github.com/torvalds/linux/blob/e93c9c99a629c61837d5a7fc2120cd2b6c70dbdd/include/uapi/linux/io_uring.h#L53-L55
- {send, recv}msg were introduced in linux v5.3: https://github.com/torvalds/linux/blob/4d856f72c10ecb060868ed10ff1b1453943fc6c8/include/uapi/linux/io_uring.h#L62-L63
- send and recv were introduced in linux v5.6: https://github.com/torvalds/linux/blob/7111951b8d4973bda27ff663f2cf18b663d15b48/include/uapi/linux/io_uring.h#L112-L113

these findings match the tags in https://unixism.net/loti/ref-liburing/submission.html .

will fallback to the aio implementation if a certain op does not exist. i understand the rationales but i think users/developers who want to use io_uring are more likely to use a newer kernel just to enjoy the new features. in my case, i'd even try out kernel 6.0 just for the zero copy send/recv. anyway, probably we can raise the bar in one years or two. as i think from the user's perspective, the difference between v5.1 and v5.3 or even v5.6 is not that large.

Reply:
just realized that read and write were added in v5.6. see https://unixism.net/loti/ref-liburing/submission.html#c.io_uring_prep_read . so we require linux 5.6 already . but i will add the machinery to detect the supported ops anyway. as i originally wanted to implement the [zero-copy transmission](https://lwn.net/Articles/879724/). in the long run, there are good chances that we support some bleeding edge ops, while user might fail to use Seastar io_uring backend just because his/her clients don't want to upgrade the kernel to linux v10.10!

Reply:
after a second thought, i think it'd be difficult to use a "downgraded" uring backend if we want to support the read and write ops. we would have to fallback to the aio backend for handling the ops in io_sink. so, can we make v5.6 the minimum linux kernel supported by the io_uring backend. the `reactor_backend_uring_v5{1,3,6}` change i just pushed is still useful, but only when we need to implement a feature only available in a newer kernel than v5.6. `IORING_OP_SENDMSG_ZC` support would be a good chance for using this mechinary. but probably not this time. i pushed it just in case. but personally, i don't really want to have this complexity at this moment.

Reply:
I don't understand. Can't we make all the code added in this series optional? Then everything beyond the basic PREAD/PWRITE/POLL is detected and optional.

Reply:
the first batch got added to linux kernel were readv, writev, fsync, read_fixed, write_fixed, poll_add, and poll_remove. but our io_uring support already uses read, write and poll_add. in which, read and write were added in linux 5.6. also, we don't use readv or writev yet. so the minimum required linux version of our io_uring support is already linux 5.6.

Reply:
I see. But we do use readv, for disk I/O.

Reply:
Note, we could use READV instead of READ. But there's no need to be so strict and we can settle on 5.6 as our base. In any case io_uring's value becomes important with sendmsg/recvmsg.

Reply:
ahh, right. we do use readv. i missed it.